autogen/agentchat/contrib/society_of_mind_agent.py

```diff
@@ -0,0 +1,108 @@
+        default_auto_reply: Optional[Union[str, Dict, None]] = "",
+    ):
+        super().__init__(
+            name=name,
+            system_message="",
+            is_termination_msg=is_termination_msg,
+            max_consecutive_auto_reply=max_consecutive_auto_reply,
+            human_input_mode=human_input_mode,
+            function_map=function_map,
+            code_execution_config=code_execution_config,
+            llm_config=llm_config,
+            default_auto_reply=default_auto_reply,
+        )
+
+        self.update_chat_manager(chat_manager)
+        self.response_preparer = response_preparer
+
+        self._reply_func_list = []
+        self.register_reply([Agent, None], SocietyOfMindAgent.generate_inner_monologue_reply)
+        self.register_reply([Agent, None], ConversableAgent.generate_code_execution_reply)
+        self.register_reply([Agent, None], ConversableAgent.generate_function_call_reply)
+        self.register_reply([Agent, None], ConversableAgent.check_termination_and_human_reply)
+
+    @property
+    def chat_manager(self) -> Union[GroupChatManager, None]:
+        """Return the group chat manager."""
+        return self._chat_manager
+
+    def update_chat_manager(self, chat_manager: Union[GroupChatManager, None]):
+        """Update the chat manager.
+
+        Args:
+            chat_manager (GroupChatManager): the group chat manager
+        """
+        self._chat_manager = chat_manager
+
+        # Awkward, but due to object cloning, there's no better way to do this
+        # Read the GroupChat object from the callback
+        self._group_chat = None
+        if self._chat_manager is not None:
+            for item in self._chat_manager._reply_func_list:
+                if isinstance(item["config"], GroupChat):
+                    self._group_chat = item["config"]
+                    break
+
+    def generate_inner_monologue_reply(
+        self,
+        messages: Optional[List[Dict]] = None,
+        sender: Optional[Agent] = None,
+        config: Optional[OpenAIWrapper] = None,
+    ) -> Tuple[bool, Union[str, Dict, None]]:
+        """Generate a reply by running the group chat"""
+        if self.chat_manager is None:
+            return False, None
+        if messages is None:
+            messages = self._oai_messages[sender]
+
+        # We want to clear the inner monolgue, keeping only the exteranl chat for context.
+        # Reset all the counters and histories, then populate agents with necesssary context from the extennal chat
+        self.chat_manager.reset()
+        self.update_chat_manager(self.chat_manager)
+
+        external_history = []
+        if len(messages) > 1:
+            external_history = messages[0 : len(messages) - 1]
+
+        for agent in self._group_chat.agents:
+            agent.reset()
+            for message in external_history:
+                self.chat_manager.send(message, agent, request_reply=False, silent=True)
```

Comment:
We should set the "name" field of the message before sending it, like what's done in group chat manager.

Reply:
I think I saw the names were already in there when I looked. It's possible this was because the responses in the tests were generated from prior invocations that included the group_chat inner monologue. Let me dig into this more.

Reply:
Ok, I believe names are set correctly now. @sonichi